variables.tf

```diff
@@ -2,40 +2,34 @@ variable "topic" {
   description = "Name of the Hedwig topic (should contain alphanumeric and dashes only by convention); unique across your infra"
 }
 
-variable "enable_firehose_all_messages" {
-  description = "Should all messages published to this topic be firehosed into Cloud Storage"
-  type        = bool
-  default     = false
-}
+variable "firehose_config" {
+  description = "Variable firehose_config describes how to \"firehose\"—read as, \"save\"—Hedwig messages to Google Cloud Storage (GCS). Under the hood, firehose_config uses the cloud_storage_config google_pubsub_subscription Terraform block; for an example, see https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/pubsub_subscription#example-usage---pubsub-subscription-push-cloudstorage-avro page."
+  type = object({
+    # Declares the bucket to firehose messages to.
+    # Gotcha: Your bucket should already exist with complicated IAM permissions—see https://cloud.google.com/pubsub/docs/create-cloudstorage-subscription#assign_roles_cloudstorage page.
+    # Gotcha: You can't actually set bucket to null; but this null default is just to express the fact that firehose_config variable is optional and if you do set firehose_config.enabled = true; but don't set bucket, then the plan will fail.
```

Comment:
Hm why not make this object optional so the presence of it signifies whether firehose is enabled or not? That way you don't need this gotcha or the "enabled" field?

Reply:
Yeah I like strongly typed objects as much as feasible and this gotcha can be avoided. Would making default as null solve your problem?

Reply:
That works—works around the bucket gotcha! And also achieves the following:
1. Make firehose_config object _optional_ by defaulting to null.
1. Declare the bucket as a _required_ string field.

Guess: I must have overlooked that possibility where you default your object the null; but the required field still won't error—surprise; but it works.